Canopy tutorials

Espeer5 opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
<!--- One sentence to describe the purpose of this PR, refer to any linked issues:
#14 -- this will link to issue 14
Closes #2 -- this will automatically close issue 2 on PR merge
--> 
Add canopy standalone and coupled hydraulics tutorials to ClimaLSM/docs. See issues #240, #255

## To-do
<!---  list of proposed tasks in the PR, move to "Content" on completion 
- Proposed task
-->
- Add links to dynamics docs at locations marked with TODO once dynamic docs are live.

## Content
<!---  specific tasks that are currently complete 
- Solution implemented
-->
- Two usage tutorials included in docs/tutorials/Canopy

<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [x] I have read and checked the items on the review checklist.

AlexisRenchon reviewed (approved):
Hi @Espeer5, your tutorial looks great, thanks!! Of course the obvious thing to fix is the dynamic docs link, but I imagine you are waiting for the actual merged into main docs link, which doesn't exist yet. 
I think what you can do for now is give the preview dynamic docs link, and add a comment on the dynamicdocs PR mentioning to correct those links when merged.

kmdeck commented:
@Espeer5 I will review today! 

Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# (https://doi.org/10.5194/gmd-14-6741-2021) for details on the site and
+# canopy parameters.
+
+# The canopy biophysics model in ClimaLSM couples a photosynthesis model with a 
```
couples -> combines or integrates

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# canopy parameters.
+
+# The canopy biophysics model in ClimaLSM couples a photosynthesis model with a 
+# canopy radiative transfer scheme and plant hydraulics model, placing them 
```
also a stomatal conductance model

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# The canopy biophysics model in ClimaLSM couples a photosynthesis model with a 
+# canopy radiative transfer scheme and plant hydraulics model, placing them 
+# under either prescribed or simulated atmospheric and radiative flux 
+# conditions. 
```
"under either prescribed or simulated (as in a full Earth System Model) atmospheric and radiative flux conditions."

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# Beer-Lambert law, in which the intensity of light absorbed is a negative 
+# exponential function of depth in the canopy and an exinction coefficient
+# determined by optical depth. For details on the Beer-Lambert model see 
+# (TODO: Dynamic Doc link).
```
could we remove the TODO comments? We can add the sentence back in when the dynamic docs are live.

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# leaf-level photosynthesis. For details on the Farquar model see 
+# (TODO: Dynamic doc link).
+
+# The plant hydraulics model in CliMA LSM uses a Van Genuchten relationship with 
```
"The plant hydraulics model in ClimaLSM solves for the water content within bulk root-stem-canopy system using Richards equation discretized into an arbitrary number of layers. The water content is related to the water potential using a retention curve relationship, and the water potential is used to simulate the effect moisture stress has on transpiration and GPP."

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# the canopy to atmospheric or soil physics models, so we choose a 
+# ['CanopyModel'](@ref ClimaLSM.Canopy.CanopyModel). From the linked 
+# documentation, we can see that we need to provide shared parameters, a domain, 
+# a radiative transfer model, photosynthesis model, plant hydraulics model, and 
```
and a stomatal conductance model

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+nelements = 10
+zmin = FT(-2)
+zmax = FT(0)
+land_domain = Point(; z_sfc = FT(0.0))
```
maybe explain that for 2D simulations, this would change, and can link to the ClimaLSM Domains documentation?

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# Create the component conductivity and retention models of the hydraulics 
+# model:
+
+K_sat_plant = 1.8e-8
```
FT()

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+# Setup the domain for the PlantHydraulics model. We will use 1 stem 
+# compartment with a single leaf compartment stacked on top.
+
+n_stem = Int64(1)
```
already defined above?

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+update_aux! = make_update_aux(canopy)
+update_aux!(p, Y, 0.0);
+
+# Select a time range to perform time stepping over, and a dt. OrdinaryDiffEq 
```
We are now using ClimaTimeSteppers - which does not support adaptive stepping.

kmdeck commented on docs/tutorials/Canopy/canopy_tutorial.jl:
```diff
@@ -0,0 +1,365 @@
+    plt2,
+    daily,
+    T,
+    label = "Modeled T",
```
"Model" to be consistent with GPP plot?

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+# # Coupling the CliMA Canopy and Soil Hydraulics Models
+
+# In the previous tutorial, we demonstrated how to run the canopy model in 
```
link to prev tutorial

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+
+
+# In ClimaLSM, the coupling of the canopy and Richard's models is done by 
+# pairing the root extraction of the canopy hydraulics model to a ['source term'](@ref ClimaLSM.Soil.AbstractSoilSource) 
```
"root extraction of the canopy hydraulics model, which acts as a boundary flux for the plant system, with a ['source term']..."

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+
+# For our soil model, we will choose a 
+# ["Richard's Model"](@ref ClimaLSM.Soil.RichardsModel) and set up all the 
+# necessary arguments. See the tutorial on Richard's Model for a more detailed 
```
perhaps link to the Richards tutorial?

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+
+# Begin by declaring the component types of the canopy model. Unlike in the 
+# previous tutorial, collect the arguments to each component into tuples and do 
+# not instantiate the component models yet.
```
We might note that the SoilPlantHydrologyModel constructor will take these arguments and create the CanopyModel and RichardsModel instances internally.

kmdeck replied:
(we can also note that this is done this way because the constructor has to enforce certain constraints on the models - e.g. that the root extraction is handled consistently between the two models - and in order to do this, the model creation must be done internally.)

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+);
+
+# Now we can initialize the state vectors and model coordinates, and initialize 
+# the explicit/implicit tendencies as usual.
```
Perhaps note that time-stepping Richards equation in the soil is done via an implicit method, while the canopy model can be stepped explicitly, so we use an IMEX (implicit-explicit) scheme for the combined model.

kmdeck commented on docs/tutorials/Canopy/soil_plant_hydrology_tutorial.jl:
```diff
@@ -0,0 +1,457 @@
+# Now initialize the auxiliary variables for the combined soil and plant model.
+
+update_aux! = make_update_aux(land)
+update_aux!(p, Y, 0.0);
```
replace 0.0 with t0

kmdeck replied:
(though you define t0 lower down)

kmdeck reviewed (commented):
hi Edward, it looks good! I left a number of comments but the tutorials are in great shape.

kmdeck commented:
> @Espeer5 I will review today!
> 
> Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.

I can help you with this on Monday if you want - I usually make a backup branch before doing it just in case (rebasing changes the commit history, so there is the potential for lost code if something goes wrong).

Espeer5 commented:
> > @Espeer5 I will review today!
> > Just a note: before merging a pull request, after addressing review comments, we squash and rebase (to only merge a single commit with all the changes; it's a bit cleaner that way). I usually follow these directions: https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request.
> 
> I can help you with this on Monday if you want - I usually make a backup branch before doing it just in case (rebasing changes the commit history, so there is the potential for lost code if something goes wrong).

That's alright, I went ahead and did it this morning. I also have implemented the changes you mentioned in your comments.

kmdeck reviewed (approved):
Thanks for addressing the review comments!. I just noticed one change in soil_energy_hydrology tutorial which I think crept in accidentally. All set to merge then after reverting/squashing!

to merge, you comment "bors r+" on the conversation of the PR, which triggers the bors application. bors runs all the tests again and once they pass will merge the PR into main.

to cancel, for any reason, you can do "bors cancel" or "bors r-".

kmdeck commented:
bors r+

kmdeck commented:
@Espeer5 not sure why buildkite failed in the bors run - it passed in github actions, and your changes didnt change the experiments folder at all. let's see if it passes this time..
